File: javascript/dynamic-programming/tabulation/canConstruct.js
// Write a function "canConstruct(target, wordBank)" that accepts a target string and an array of strings.
// The function should return a boolean indicating whether or not the "target" can be constructed by concatenating elements of the "wordBank" array.

function canConstruct(target, wordBank) {
   const table = Array(target.length + 1).fill(false)
    table[0] = true

    for (let i = 0; i <= target.length; i++) {
        if (table[i] === true) {
            for (let word of wordBank) {
                if (target.slice(i, i + word.length) === word) {
                    table[i + word.length] = true
                }
            }
        }
    }

    return table[target.length]
}

// testing
console.log(canConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"]))
// expected output: true
console.log(canConstruct("eeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "f"]))
// expected output: true

File: python/dynamic-programming/tabulation/canConstruct.py
# Write a function "canConstruct(target, wordBank)" that accepts a target string and an array of strings.
# The function should return a boolean indicating whether or not the "target" can be constructed by concatenating elements of the "wordBank" array.

def canConstruct(target: str, wordBank: list) -> bool:
    table = [False] * (len(target) + 1)
    table[0] = True

    for i in range(len(target)):
        if table[i]:
            for word in wordBank:
                if target[i : i + len(word)] == word:
                    table[i + len(word)] = True
    
    return table[len(target)]

# testing
print(canConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"]))
# expected output: True
print(canConstruct("eeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "f"]))
# expected output: True